File: day5/part2.go
package main

import (
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"strconv"
	"strings"
	"sync"
	"sync/atomic"
	"time"
)

func main() {
	currDir, _ := os.Getwd()
	os.Chdir(currDir)

	content, err := ioutil.ReadFile("input.txt")
	if err != nil {
		fmt.Println("Error reading input file:", err)
		return
	}

	lines := strings.Split(string(content), "\n")

	seeds := strings.Split(lines[0], ": ")[1]
	seedLine := strings.Fields(seeds)
	seedRanges := make([]int, 0)
	for _, s := range seedLine {
		if s != "" {
			num, _ := strconv.Atoi(s)
			seedRanges = append(seedRanges, num)
		}
	}

	v2(lines, seedRanges) // Brute force locations
	// v1(lines, seedRanges) // Brute force seeds .... LOL
}

func v2(lines []string, seedRanges []int) {
	lowest := 999999999
	index := lowest
	for {
		index -= 100000
		if index < 0 {
			index = lowest + 1
			break
		}
		valid := checkIfLocationIsValid(index, lines)
		if valid {
			lowest = index
		}
		fmt.Println("Lowest:", lowest, "Index:", index)
	}
	multiplier := 1
	streak := 0
	for {
		index -= multiplier
		if index < 0 {
			break
		}
		valid := checkIfLocationIsValid(index, lines)
		if valid {
			lowest = index
			streak = 0
			multiplier = 1
		} else {
			streak++
			if streak > 1000 && multiplier < 50000 {
				multiplier += 1
			}
		}
		fmt.Println("Lowest:", lowest, "Index:", index)
	}
}

func checkIfLocationIsValid(location int, lines []string) bool {
	index := len(lines) - 1

	allLocations := false
	locationToHumidity := make(map[int][2]int)
	for !allLocations {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allLocations = true
			index -= 2
			break
		}
		l, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		locationToHumidity[l] = [2]int{r, d - l}
		index--
	}
	humidity := location
	for k, v := range locationToHumidity {
		if humidity >= k && humidity <= k+v[0] {
			humidity += v[1]
			break
		}
	}
	// fmt.Println("-- Humidity:", humidity)

	allHumidities := false
	humidityToTemperature := make(map[int][2]int)
	for !allHumidities {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allHumidities = true
			index -= 2
			break
		}
		h, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		humidityToTemperature[h] = [2]int{r, d - h}
		index--
	}
	temperature := humidity
	for k, v := range humidityToTemperature {
		if humidity >= k && humidity <= k+v[0] {
			temperature += v[1]
			break
		}
	}
	// fmt.Println("-- Temperature:", temperature)

	allTemperatures := false
	temperatureToLight := make(map[int][2]int)
	for !allTemperatures {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allTemperatures = true
			index -= 2
			break
		}
		t, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		temperatureToLight[t] = [2]int{r, d - t}
		index--
	}
	light := temperature
	for k, v := range temperatureToLight {
		if temperature >= k && temperature <= k+v[0] {
			light += v[1]
			break
		}
	}
	// fmt.Println("-- Light:", light)

	allLights := false
	lightToWater := make(map[int][2]int)
	for !allLights {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allLights = true
			index -= 2
			break
		}
		l, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		lightToWater[l] = [2]int{r, d - l}
		index--
	}
	water := light
	for k, v := range lightToWater {
		if light >= k && light <= k+v[0] {
			water += v[1]
			break
		}
	}
	// fmt.Println("-- Water:", water)

	allWaters := false
	waterToFertilizer := make(map[int][2]int)
	for !allWaters {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allWaters = true
			index -= 2
			break
		}
		w, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		waterToFertilizer[w] = [2]int{r, d - w}
		index--
	}
	fertilizer := water
	for k, v := range waterToFertilizer {
		if water >= k && water <= k+v[0] {
			fertilizer += v[1]
			break
		}
	}
	// fmt.Println("-- Fertilizer:", fertilizer)

	allFertilizers := false
	fertilizerToSoil := make(map[int][2]int)
	for !allFertilizers {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allFertilizers = true
			index -= 2
			break
		}
		f, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		fertilizerToSoil[f] = [2]int{r, d - f}
		index--
	}
	soil := fertilizer
	for k, v := range fertilizerToSoil {
		if fertilizer >= k && fertilizer <= k+v[0] {
			soil += v[1]
			break
		}
	}
	// fmt.Println("-- Soil:", soil)

	allSoils := false
	soilToSeed := make(map[int][2]int)
	for !allSoils {
		line := strings.Fields(lines[index])
		if len(line) != 3 {
			allSoils = true
			index -= 2
			break
		}
		s, _ := strconv.Atoi(line[0])
		d, _ := strconv.Atoi(line[1])
		r, _ := strconv.Atoi(line[2])
		soilToSeed[s] = [2]int{r, d - s}
		index--
	}
	seed := soil
	for k, v := range soilToSeed {
		if soil >= k && soil <= k+v[0] {
			seed += v[1]
			break
		}
	}
	// fmt.Println("-- Seed:", seed)

	// check if seed is in one of the seed ranges
	seeds := strings.Split(lines[0], ": ")[1]
	seedLine := strings.Fields(seeds)
	seedRanges := make([]int, 0)
	for _, s := range seedLine {
		if s != "" {
			num, _ := strconv.Atoi(s)
			seedRanges = append(seedRanges, num)
		}
	}

	for i := 0; i < len(seedRanges); i += 2 {
		if seed >= seedRanges[i] && seed <= seedRanges[i+1]+seedRanges[i] {
			// fmt.Println("-- Seed is in range:", seedRanges[i], "-", seedRanges[i+1]+seedRanges[i])
			return true
		}
	}

	// fmt.Println("-- Seed is not in range")
	return false
}

func v1(lines []string, seedRanges []int) {
	splits := 10000000

	splitRanges := make([]int, 0)
	for i := 0; i < len(seedRanges); i += 2 {
		inc := seedRanges[i+1] / splits
		rem := seedRanges[i+1] % splits
		splitRanges = append(splitRanges, seedRanges[i])
		splitRanges = append(splitRanges, inc)
		for j := 0; j < splits-1; j++ {
			splitRanges = append(splitRanges, seedRanges[i]+inc*(j+1))
			splitRanges = append(splitRanges, inc)
		}
		splitRanges = append(splitRanges, seedRanges[i]+inc*splits)
		splitRanges = append(splitRanges, rem)
	}

	// print the number of seeds in each range
	// for i := 0; i < len(splitRanges); i += 2 {
	// 	fmt.Println("Range:", splitRanges[i], "-", splitRanges[i+1]+splitRanges[i], "has", splitRanges[i+1], "seeds")
	// }

	// return

	// lowestList := make([]int, 0)
	var counter int64 = 0
	var globalLowest int64 = 1500948298
	var wg sync.WaitGroup
	fmt.Println("Waiting for workers to finish...")

	// periodically print the number of workers still running
	// go func() {
	// 	for {
	// 		fmt.Println("Workers remaining:", atomic.LoadInt64(&counter))
	// 		time.Sleep(1 * time.Second)
	// 	}
	// }()

	for i := 0; i < len(splitRanges); i += 2 {
		wg.Add(1)
		go func(seedRanges []int, i int, lines []string) {
			atomic.AddInt64(&counter, 1)
			defer wg.Done()
			findLowest(splitRanges, i, lines, &globalLowest)
			atomic.AddInt64(&counter, -1)
		}(splitRanges, i, lines)
	}

	wg.Wait()
	time.Sleep(1 * time.Second)

	fmt.Println("Lowest location:", globalLowest)
}

func findLowest(seedRanges []int, i int, lines []string, globalLowest *int64) {
	// log.Println("Checking seed range:", seedRanges[i], "-", seedRanges[i+1]+seedRanges[i])
	// counter := 0
	for seed := seedRanges[i]; seed <= seedRanges[i]+seedRanges[i+1]; seed++ {
		// log.Println("Checking seed:", seed)
		// counter++
		// if counter%1000 == 0 {
		// 	log.Println("Checked", counter, "seeds")
		// }
		index := 3

		allSoils := false
		seedToSoil := make(map[int][2]int)
		for !allSoils {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allSoils = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			seedToSoil[lookup] = [2]int{r, dest - lookup}
			index++
		}
		soil := seed
		for k, v := range seedToSoil {
			if seed >= k && seed <= k+v[0] {
				soil += v[1]
				break
			}
		}

		allFertilizers := false
		soilToFertilizer := make(map[int][2]int)
		for !allFertilizers {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allFertilizers = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			soilToFertilizer[lookup] = [2]int{r, dest - lookup}
			index++
		}
		fertilizer := soil
		for k, v := range soilToFertilizer {
			if soil >= k && soil <= k+v[0] {
				fertilizer += v[1]
				break
			}
		}

		allWaters := false
		fertilizerToWater := make(map[int][2]int)
		for !allWaters {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allWaters = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			fertilizerToWater[lookup] = [2]int{r, dest - lookup}
			index++
		}
		water := fertilizer
		for k, v := range fertilizerToWater {
			if fertilizer >= k && fertilizer <= k+v[0] {
				water += v[1]
				break
			}
		}

		allLights := false
		waterToLight := make(map[int][2]int)
		for !allLights {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allLights = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			waterToLight[lookup] = [2]int{r, dest - lookup}
			index++
		}
		light := water
		for k, v := range waterToLight {
			if water >= k && water <= k+v[0] {
				light += v[1]
				break
			}
		}

		allTemperatures := false
		lightToTemperature := make(map[int][2]int)
		for !allTemperatures {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allTemperatures = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			lightToTemperature[lookup] = [2]int{r, dest - lookup}
			index++
		}
		temperature := light
		for k, v := range lightToTemperature {
			if light >= k && light <= k+v[0] {
				temperature += v[1]
				break
			}
		}

		allHumidities := false
		temperatureToHumidity := make(map[int][2]int)
		for !allHumidities {
			line := strings.Fields(lines[index])
			if len(line) != 3 {
				allHumidities = true
				index += 2
				break
			}
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			temperatureToHumidity[lookup] = [2]int{r, dest - lookup}
			index++
		}
		humidity := temperature
		for k, v := range temperatureToHumidity {
			if temperature >= k && temperature <= k+v[0] {
				humidity += v[1]
				break
			}
		}

		humidityToLocation := make(map[int][2]int)
		for index < len(lines) {
			line := strings.Fields(lines[index])
			dest, _ := strconv.Atoi(line[0])
			lookup, _ := strconv.Atoi(line[1])
			r, _ := strconv.Atoi(line[2])
			humidityToLocation[lookup] = [2]int{r, dest - lookup}
			index++
		}
		location := humidity
		for k, v := range humidityToLocation {
			if humidity >= k && humidity <= k+v[0] {
				location += v[1]
				break
			}
		}

		if location < int(atomic.LoadInt64(globalLowest)) {
			atomic.StoreInt64(globalLowest, int64(location))
			log.Println("New lowest location:", location)
		}
	}
}

File: 2023/day05/part2.py
import os
curr_dir = os.path.dirname(os.path.abspath(__file__))
os.chdir(curr_dir)

with open("input.txt") as f:
    lines = f.readlines()

lowest = 99999999999999
seeds = []
seed_line = lines[0].split(': ')[1].replace('\n', '').split(' ')

seed_ranges = [int(s) for s in seed_line if s != '']
for i in range(0, len(seed_ranges), 2):
    print(f"Checking {seed_ranges[i+1]} seeds starting at {seed_ranges[i]}")
    counter = 0
    for seed in range(seed_ranges[i], seed_ranges[i]+seed_ranges[i+1]):
        counter += 1
        if counter % 10000 == 0:
            print("Checked", counter, "seeds...")
        index = 3

        # read in soil map
        all_soils = False
        seed_to_soil = {}
        while not all_soils:
            if lines[index] == '\n':
                all_soils = True
                index += 2
                break

            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            seed_to_soil[lookup] = r, dest-lookup
            index += 1

        # apply soil map to seed
        soil = seed
        for k, v in seed_to_soil.items():
            if seed >= k and seed <= k+v[0]:
                soil += v[1]
                break
        # free up memory
        del seed_to_soil

        # read in fertilizer map
        all_fertilizers = False
        soil_to_fertilizer = {}
        while not all_fertilizers:
            if lines[index] == '\n':
                all_fertilizers = True
                index += 2
                break
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            soil_to_fertilizer[lookup] = r, dest-lookup
            index += 1
        # apply fertilizer map to soil
        fertilizer = soil
        for k, v in soil_to_fertilizer.items():
            if soil >= k and soil <= k+v[0]:
                fertilizer += v[1]
                break
        # free up memory
        del soil_to_fertilizer

        # read in water map
        all_waters = False
        fertilizer_to_water = {}
        while not all_waters:
            if lines[index] == '\n':
                all_waters = True
                index += 2
                break
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            fertilizer_to_water[lookup] = r, dest-lookup
            index += 1

        # apply water map to fertilizers
        water = fertilizer
        for k, v in fertilizer_to_water.items():
            if fertilizer >= k and fertilizer <= k+v[0]:
                water += v[1]
                break
        # free up memory
        del fertilizer_to_water

        # read in light map
        all_lights = False
        water_to_light = {}
        while not all_lights:
            if lines[index] == '\n':
                all_lights = True
                index += 2
                break
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            water_to_light[lookup] = r, dest-lookup
            index += 1
        # apply light map to waters
        light = water
        for k, v in water_to_light.items():
            if water >= k and water <= k+v[0]:
                light += v[1]
                break
        # free up memory
        del water_to_light

        # read in temperature map
        all_temperatures = False
        light_to_temperature = {}
        while not all_temperatures:
            if lines[index] == '\n':
                all_temperatures = True
                index += 2
                break
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            light_to_temperature[lookup] = r, dest-lookup
            index += 1
        # apply temperature map to lights
        temperature = light
        for k, v in light_to_temperature.items():
            if light >= k and light <= k+v[0]:
                temperature += v[1]
                break
        # free up memory
        del light_to_temperature

        # read in humidity map
        all_humidities = False
        temperature_to_humidity = {}
        while not all_humidities:
            if lines[index] == '\n':
                all_humidities = True
                index += 2
                break
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            temperature_to_humidity[lookup] = r, dest-lookup
            index += 1
        # apply humidity map to temperatures
        humidity = temperature
        for k, v in temperature_to_humidity.items():
            if temperature >= k and temperature <= k+v[0]:
                humidity += v[1]
                break
        # free up memory
        del temperature_to_humidity

        # read in location map
        humidity_to_location = {}
        while index < len(lines):
            line = lines[index].replace('\n', '').split(' ')
            dest = int(line[0])
            lookup = int(line[1])
            r = int(line[2])
            humidity_to_location[lookup] = r, dest-lookup
            index += 1
        # apply location map to humidities
        location = humidity
        for k, v in humidity_to_location.items():
            if humidity >= k and humidity <= k+v[0]:
                location += v[1]
                break
        # free up memory
        del humidity_to_location

        if location < lowest:
            lowest = location

# print the lowest location
print("Lowest Location:", lowest)

# debug
# for i in range(len(seeds)):
#     print(seeds[i], soils[i], fertilizers[i], waters[i], lights[i], temperatures[i], humidities[i], locations[i])


# first number is destination
# second number is lookup
# third number is the range

# walk each seed through the maps
    # check if the seed is in one of the (lookup to lookup+range) seed-to-soil maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the soil is the same as the seed
    # check if the soil is in one of the (lookup to lookup+range) soil-to-fertilizer maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the fertilizer is the same as the soil
    # check if the fertilizer is in one of the (lookup to lookup+range) fertilizer-to-water maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the water is the same as the fertilizer
    # check if the water is in one of the (lookup to lookup+range) water-to-light maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the light is the same as the water
    # check if the light is in one of the (lookup to lookup+range) light-to-temperature maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the temperature is the same as the light
    # check if the temperature is in one of the (lookup to lookup+range) temperature-to-humidity maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the humidity is the same as the temperature
    # check if the humidity is in one of the (lookup to lookup+range) humidity-to-location maps
        # if it is, add the difference of (destination-lookup) to the location
        # if it is not, the location is the same as the humidity

    # record the location of the seed in a dictionary with the seed as the key and the location as the value    

# after all seeds have been walked through the maps, print the lowest location